File: backend/src/app/modules/admin/services/ip_block_service.py
from __future__ import annotations

from datetime import datetime
from ipaddress import IPv4Address, IPv6Address, ip_address

from sqlalchemy import func, or_, select
from sqlalchemy.orm import Session

from app.core.time import utc_now
from app.modules.accounts.models.user import User
from app.modules.admin.models.ip_block import IpBlock
from app.modules.admin.schemas.ip_block import IpBlockCreate, IpBlockRead, IpBlockSummary


class IpBlockError(ValueError):
    pass


def normalize_ip_address(value: str | None) -> str | None:
    raw = (value or "").strip()
    if not raw:
        return None
    try:
        parsed = ip_address(raw)
    except ValueError as exc:
        raise IpBlockError("Enter a valid individual IPv4 or IPv6 address.") from exc
    if isinstance(parsed, IPv6Address) and parsed.ipv4_mapped:
        parsed = parsed.ipv4_mapped
    return parsed.compressed


def validate_blockable_ip(value: str) -> str:
    normalized = normalize_ip_address(value)
    if normalized is None:
        raise IpBlockError("Enter an IP address.")
    parsed: IPv4Address | IPv6Address = ip_address(normalized)
    if parsed.is_loopback or parsed.is_unspecified or parsed.is_multicast:
        raise IpBlockError("Loopback, unspecified and multicast addresses cannot be blocked.")
    return normalized


def _is_active(row: IpBlock, now: datetime | None = None) -> bool:
    now = now or utc_now()
    return row.unblocked_at is None and (row.expires_at is None or row.expires_at > now)


def to_read(row: IpBlock, now: datetime | None = None) -> IpBlockRead:
    now = now or utc_now()
    expired = row.unblocked_at is None and row.expires_at is not None and row.expires_at <= now
    return IpBlockRead(
        id=row.id,
        ip_address=row.ip_address,
        reason=row.reason,
        notes=row.notes,
        created_at=row.created_at,
        created_by_user_id=row.created_by_user_id,
        created_by_username=row.created_by_username,
        expires_at=row.expires_at,
        unblocked_at=row.unblocked_at,
        unblocked_by_user_id=row.unblocked_by_user_id,
        unblocked_by_username=row.unblocked_by_username,
        unblock_reason=row.unblock_reason,
        is_active=_is_active(row, now),
        is_temporary=row.expires_at is not None,
        is_expired=expired,
    )


def find_active_ip_block(db: Session, value: str | None) -> IpBlock | None:
    try:
        normalized = normalize_ip_address(value)
    except IpBlockError:
        return None
    if normalized is None:
        return None
    now = utc_now()
    return db.scalar(
        select(IpBlock)
        .where(
            IpBlock.ip_address == normalized,
            IpBlock.unblocked_at.is_(None),
            or_(IpBlock.expires_at.is_(None), IpBlock.expires_at > now),
        )
        .order_by(IpBlock.created_at.desc(), IpBlock.id.desc())
        .limit(1)
    )


def create_ip_block(db: Session, *, actor: User, payload: IpBlockCreate) -> IpBlockRead:
    normalized = validate_blockable_ip(payload.ip_address)
    if len(payload.reason.strip()) < 3:
        raise IpBlockError("Enter a reason with at least three characters.")
    now = utc_now()
    if payload.expires_at is not None and payload.expires_at <= now:
        raise IpBlockError("Expiration must be in the future.")
    if find_active_ip_block(db, normalized) is not None:
        raise IpBlockError("This IP address is already blocked.")
    row = IpBlock(
        ip_address=normalized,
        reason=payload.reason.strip(),
        notes=payload.notes,
        created_by_user_id=actor.id,
        created_by_username=actor.username,
        expires_at=payload.expires_at,
    )
    db.add(row)
    db.commit()
    db.refresh(row)
    return to_read(row)


def unblock_ip_block(db: Session, *, block_id: int, actor: User, reason: str | None = None) -> IpBlockRead:
    row = db.get(IpBlock, block_id)
    if row is None:
        raise IpBlockError("IP block not found.")
    if not _is_active(row):
        raise IpBlockError("This IP block is no longer active.")
    row.unblocked_at = utc_now()
    row.unblocked_by_user_id = actor.id
    row.unblocked_by_username = actor.username
    row.unblock_reason = (reason or "").strip() or None
    db.commit()
    db.refresh(row)
    return to_read(row)


def list_ip_blocks(
    db: Session,
    *,
    status: str = "active",
    search: str | None = None,
    limit: int = 200,
) -> list[IpBlockRead]:
    now = utc_now()
    query = select(IpBlock)
    if status == "active":
        query = query.where(
            IpBlock.unblocked_at.is_(None),
            or_(IpBlock.expires_at.is_(None), IpBlock.expires_at > now),
        )
    elif status == "expired":
        query = query.where(
            IpBlock.unblocked_at.is_(None),
            IpBlock.expires_at.is_not(None),
            IpBlock.expires_at <= now,
        )
    elif status == "unblocked":
        query = query.where(IpBlock.unblocked_at.is_not(None))
    elif status != "all":
        raise IpBlockError("Invalid IP block status filter.")
    if search:
        needle = search.strip()
        query = query.where(
            or_(
                IpBlock.ip_address.contains(needle),
                IpBlock.reason.contains(needle),
                IpBlock.created_by_username.contains(needle),
            )
        )
    rows = db.scalars(query.order_by(IpBlock.created_at.desc(), IpBlock.id.desc()).limit(limit)).all()
    return [to_read(row, now) for row in rows]


def ip_block_summary(db: Session) -> IpBlockSummary:
    now = utc_now()
    total = int(db.scalar(select(func.count(IpBlock.id))) or 0)
    active_filter = (
        IpBlock.unblocked_at.is_(None),
        or_(IpBlock.expires_at.is_(None), IpBlock.expires_at > now),
    )
    active = int(db.scalar(select(func.count(IpBlock.id)).where(*active_filter)) or 0)
    permanent = int(
        db.scalar(select(func.count(IpBlock.id)).where(*active_filter, IpBlock.expires_at.is_(None))) or 0
    )
    temporary = int(
        db.scalar(select(func.count(IpBlock.id)).where(*active_filter, IpBlock.expires_at.is_not(None))) or 0
    )
    expired = int(
        db.scalar(
            select(func.count(IpBlock.id)).where(
                IpBlock.unblocked_at.is_(None),
                IpBlock.expires_at.is_not(None),
                IpBlock.expires_at <= now,
            )
        )
        or 0
    )
    unblocked = int(db.scalar(select(func.count(IpBlock.id)).where(IpBlock.unblocked_at.is_not(None))) or 0)
    return IpBlockSummary(
        total=total,
        active=active,
        permanent=permanent,
        temporary=temporary,
        expired=expired,
        unblocked=unblocked,
    )

File: spring-api/src/main/java/eu/royalblackwater/api/securityops/IpBlockService.java
package eu.royalblackwater.api.securityops;

import static eu.royalblackwater.api.persistence.RowValues.*;
import static org.springframework.http.HttpStatus.CONFLICT;
import static org.springframework.http.HttpStatus.NOT_FOUND;

import eu.royalblackwater.api.audit.AuditService;
import eu.royalblackwater.api.contract.IpBlockCreate;
import eu.royalblackwater.api.contract.IpBlockRead;
import eu.royalblackwater.api.contract.IpBlockSummary;
import eu.royalblackwater.api.contract.IpBlockUnblock;
import eu.royalblackwater.api.persistence.JdbcQueryService;
import eu.royalblackwater.api.persistence.SqlParameters;
import eu.royalblackwater.api.security.AuthenticatedUser;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.time.Clock;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Service
public class IpBlockService {
    private final JdbcQueryService jdbc;
    private final AuditService audit;
    private final Clock clock;
    public IpBlockService(JdbcQueryService jdbc,AuditService audit,Clock clock){this.jdbc=jdbc;this.audit=audit;this.clock=clock;}

    @Transactional(readOnly=true)
    public boolean isBlocked(String rawIp){
        String ip=normalizeIp(rawIp,false);
        if(ip==null) return false;
        return jdbc.count("""
                select count(*) from ip_blocks where ip_address=:ip and unblocked_at is null
                  and (expires_at is null or expires_at>:now)
                """,Map.of("ip",ip,"now",now()))>0;
    }

    @Transactional(readOnly=true)
    public List<IpBlockRead> list(String status,String search,long limit){
        StringBuilder sql=new StringBuilder("select * from ip_blocks where 1=1");
        Map<String,Object> params=new LinkedHashMap<>();
        String normalized=status==null?"active":status.strip().toLowerCase();
        switch(normalized){
            case "active" -> sql.append(" and unblocked_at is null and (expires_at is null or expires_at>:now)");
            case "expired" -> sql.append(" and unblocked_at is null and expires_at<=:now");
            case "unblocked" -> sql.append(" and unblocked_at is not null");
            case "all" -> { }
            default -> throw new ResponseStatusException(org.springframework.http.HttpStatus.BAD_REQUEST,"Invalid IP-block status.");
        }
        if(!"all".equals(normalized)){params.put("now",now());}
        if(search!=null&&!search.isBlank()){sql.append(" and (ip_address ilike :search or reason ilike :search or coalesce(notes,'') ilike :search)");params.put("search","%"+search.strip()+"%");}
        sql.append(" order by created_at desc,id desc limit :limit");params.put("limit",Math.max(1,Math.min(1000,limit)));
        return jdbc.query(sql.toString(),params).stream().map(this::read).toList();
    }

    @Transactional(readOnly=true)
    public IpBlockSummary summary(){
        LocalDateTime now=now();
        return new IpBlockSummary(
                jdbc.count("select count(*) from ip_blocks where unblocked_at is null and (expires_at is null or expires_at>:now)",Map.of("now",now)),
                jdbc.count("select count(*) from ip_blocks where unblocked_at is null and expires_at<=:now",Map.of("now",now)),
                jdbc.count("select count(*) from ip_blocks where expires_at is null",Map.of()),
                jdbc.count("select count(*) from ip_blocks where expires_at is not null",Map.of()),
                jdbc.count("select count(*) from ip_blocks",Map.of()),
                jdbc.count("select count(*) from ip_blocks where unblocked_at is not null",Map.of()));
    }

    @Transactional
    public IpBlockRead create(AuthenticatedUser actor,IpBlockCreate input){
        String ip=normalizeIp(input.ipAddress(),true);
        if(input.expiresAt()!=null&&!input.expiresAt().isAfter(now())) throw new ResponseStatusException(CONFLICT,"Expiration must be in the future.");
        if(isBlocked(ip)) throw new ResponseStatusException(CONFLICT,"IP address is already actively blocked.");
        long id=jdbc.insertReturningId("""
                insert into ip_blocks(ip_address,reason,notes,created_at,created_by_user_id,created_by_username,expires_at)
                values(:ip,:reason,:notes,:now,:actorId,:actor,:expires) returning id
                """,SqlParameters.ofNullable("ip",ip,"reason",input.reason().strip(),"notes",blank(input.notes()),"now",now(),
                        "actorId",actor.id(),"actor",actor.username(),"expires",input.expiresAt()));
        audit.record(actor,"ip_block",id,"create","Blocked IP address "+ip,Set.of("reason","expires_at"));
        return required(id);
    }

    @Transactional
    public IpBlockRead unblock(AuthenticatedUser actor,long id,IpBlockUnblock input){
        Map<String,Object> row=jdbc.optional("select * from ip_blocks where id=:id",Map.of("id",id)).orElseThrow(()->new ResponseStatusException(NOT_FOUND,"IP block not found."));
        if(row.get("unblocked_at")!=null) return read(row);
        jdbc.update("""
                update ip_blocks set unblocked_at=:now,unblocked_by_user_id=:actorId,
                    unblocked_by_username=:actor,unblock_reason=:reason where id=:id
                """,SqlParameters.ofNullable("id",id,"now",now(),"actorId",actor.id(),"actor",actor.username(),"reason",blank(input.reason())));
        audit.record(actor,"ip_block",id,"unblock","Unblocked IP address",Set.of("unblock_reason"));
        return required(id);
    }

    private IpBlockRead required(long id){return read(jdbc.required("select * from ip_blocks where id=:id",Map.of("id",id)));}
    private IpBlockRead read(Map<String,Object> row){
        LocalDateTime expires=nullableDateTime(row,"expires_at");
        LocalDateTime unblocked=nullableDateTime(row,"unblocked_at");
        boolean expired=expires!=null&&!expires.isAfter(now());
        return new IpBlockRead(dateTime(row,"created_at"),nullableLong(row,"created_by_user_id"),requiredString(row,"created_by_username"),
                expires,longValue(row,"id"),requiredString(row,"ip_address"),unblocked==null&&!expired,expired,expires!=null,
                string(row,"notes"),requiredString(row,"reason"),string(row,"unblock_reason"),unblocked,
                nullableLong(row,"unblocked_by_user_id"),string(row,"unblocked_by_username"));
    }

    static String normalizeIp(String value,boolean required){
        if(value==null||value.isBlank()){if(required) throw new ResponseStatusException(org.springframework.http.HttpStatus.BAD_REQUEST,"IP address is required.");return null;}
        String raw=value.strip();
        if(!raw.matches("[0-9a-fA-F:.]+")) throw new ResponseStatusException(org.springframework.http.HttpStatus.BAD_REQUEST,"Only literal IPv4 or IPv6 addresses may be blocked.");
        try{return InetAddress.getByName(raw).getHostAddress();}catch(UnknownHostException exception){throw new ResponseStatusException(org.springframework.http.HttpStatus.BAD_REQUEST,"Invalid IP address.");}
    }
    private LocalDateTime now(){return LocalDateTime.ofInstant(clock.instant(),ZoneOffset.UTC);}
    private static String blank(String value){return value==null||value.isBlank()?null:value.strip();}
}
